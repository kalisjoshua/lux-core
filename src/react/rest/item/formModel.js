/**
 * @module react/rest/item/formModel
 * @memberof react/rest/item
 */

// FIXME: refactor this module into lib/

import { hasOne } from '../../../lib/has';
import { isArray } from '../../../lib/is';

function required() {
  throw new Error('No arguments passed to `formModel`.');
}

function formActionsReduce(acc, action) {
  // NOTE: #Siren-consideration: action.name='[match]-item'
  const match = (action.name.match(/^(create|delete|update|view)/) || [])[1];

  /* istanbul ignore else */
  if (match) {
    acc[match] = action;
  }

  return acc;
}

function formModel(props = required()) {
  if (!props.actions || !isArray(props.actions)) {
    throw new Error('No `actions` provided; "view-item" action is required.');
  }

  const actions = props.actions
    .reduce(formActionsReduce, {});

  if (!actions.view) {
    throw new Error('View action not provided; which would include fields[].');
  }

  const schema = actions.view.fields;

  // TODO: ensure that component errors are being handled appropriately;
  // resulting in an error page and not a blank page with JS console errors
  if (!schema) {
    throw new Error('No schema found.');
  }

  // make "field+config mapping" (below) simpler, and easier
  const fieldConfigs = (props.entities || [])
    .reduce((acc, ent) => {
      /* istanbul ignore else */
      if (hasOne('class', 'fieldConfig', ent)) {
        Object.keys(ent.properties)
          .forEach((field) => {
            acc[field] = {
              actions: ent.actions,
              entities: ent.entities,
              ...ent.properties[field],
            };
          });
      }

      return acc;
    }, {});

  return {
    ...actions,
    properties: /\/new\/?$/.test(props.path) ? {} : { ...props.properties },
    schema: schema
      // field+config mapping
      .map(field => ({ ...field, config: fieldConfigs[field.name] })),
    title: props.title,
  };
}

export default formModel;
