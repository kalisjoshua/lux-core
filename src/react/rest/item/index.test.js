jest.mock('./formModel');
import formModel from './formModel';

import React from 'react';
import renderer from 'react-test-renderer';

import registry from '../../componentRegistry';

import './index';
const Item = registry('Rest.Item');

const Form = registry('Form');

formModel.mockImplementation(() => {});

describe('Rest.Item', function () {
  it('should exist; and should be a function', function () {
    expect(typeof Item).toBe('function');
  });

  afterEach(function () {
    registry('Form', Form);
  });

  beforeEach(function () {
    // // silence console.error
    // spyOn(console, 'error');
  });

  it('should render a Form component', function () {
    registry('Form', () => (<p />));
    const attrs = {
      links: [],
      properties: {},
      title: '',
      path: '',
    };
    const component = renderer.create(
      <Item {...attrs} />
    );

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
