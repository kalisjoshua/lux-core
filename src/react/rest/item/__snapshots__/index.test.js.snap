exports[`Rest.Item should render a Form component 1`] = `<p />`;
